### PR TITLE
T164930 mail template preview rendering 

### PR DESCRIPTION
**Pull request by wiese**
A command to check and dump mail templates without actually sending them.

### CODE REVIEW

File: cli/RenderMailTemplatesCommand.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	/**
+	 * Render all templates and write them to disk to allow a comparison with an alternative data set
+	 *
+	 * @param array $testData Template names and fixture information to render these templates
+	 * @param Twig_Environment $twig The templating engine to render the templates
+	 * @param string $outputPath Path where rendered templates will be written to
+	 * @param OutputInterface $output Command output
+	 */
+	private function renderTemplates( array $testData, Twig_Environment $twig, string $outputPath, OutputInterface $output ): void {
+		foreach( $testData as $template => $templateSettings ) {
```
Comment (KaiNissen) [inline cli/RenderMailTemplatesCommand.php]: Err, isn't a space after `foreach` caught by the CS check?

File: cli/RenderMailTemplatesCommand.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+					$output->writeln( "$outputName already exists, skipping ..." );
+					continue;
+				}
+
+				try {
+					file_put_contents( $outputName, $twig->render( $template, array_merge_recursive(
+						$templateSettings['context'],
+						$templateSettings['variants'][$variantName]
+					) ) );
+				} catch( Twig_Error $e ) {
```
Comment (KaiNissen) [inline cli/RenderMailTemplatesCommand.php]: While we're at it: missing space after the `catch`.

**Comment by KaiNissen** [comment]
Looks good to me. Did we define acceptance criteria somewhere?

File: cli/RenderMailTemplatesCommand.php (hunk 1 above)
Comment (JeroenDeDauw) [inline cli/RenderMailTemplatesCommand.php]: Probably one of the disabled rules that blew up when we started using nullable types.

File: tests/Data/mail_templates.php [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	'Mail_Donation_Confirmation.txt.twig' => [
+		'context' => [
+			'greeting_generator' => $ffFactory->getGreetingGenerator(),
+			'donation' => [
+				'id' => 42,
+				'amount' => 12.34,
+				'needsModeration' => false,
+			],
+			'recipient' => [
+				'lastName' => '姜',
```
Comment (wiese) [inline tests/Data/mail_templates.php]: Objections? Not to sure what to make of this input.

File: cli/RenderMailTemplatesCommand.php (hunk 1 above)
Comment (wiese) [inline cli/RenderMailTemplatesCommand.php]: @JeroenDeDauw @KaiNissen Feel free to mend the phpcs.xml to what ever you think is hip today. Until then I lean towards "if it's not tested it does not exist" (the rule).

File: tests/Data/mail_templates.php [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		]
+	],
+
+	'Mail_Membership_Application_Confirmation.txt.twig' => [
+		'context' => [
+			'greeting_generator' => $ffFactory->getGreetingGenerator(),
+			'lastName' => "O'Reilly",
+			'salutation' => 'Herr',
+			'title' => 'Dr.',
+			'applicationId' => 23,
```
Comment (wiese) [inline tests/Data/mail_templates.php]: Nice catch! Unfortunately, AFAIK there is no sane way to extract variables used inside a template (and it's included templates) to automatically warn about redundant variable definitions. In the other direction, we thankfully have `strict-variables`, so variables which are not mocked will raise an error...

File: cli/RenderMailTemplatesCommand.php (hunk 2 above)
Comment (wiese) [inline cli/RenderMailTemplatesCommand.php]: There does not seem to be a rule for that (neither followed in the code, nor in phpcs) - 100+ matches inside the project w/o space. It's not much different from a normal function call which we also write w/o space, right?

    vagrant@vagrant:/vagrant$ ./vendor/bin/phpcs -v cli
    Registering sniffs in the FunFunFundraisingFrontend standard... DONE (62 sniffs registered)
    Creating file list... DONE (6 files in queue)
    Changing into directory /vagrant/cli/ConfigValidation
    Processing ConfigValidationException.php [PHP => 61 tokens in 13 lines]... DONE in 40ms (0 errors, 0 warnings)
    Processing SchemaLoader.php [PHP => 284 tokens in 37 lines]... DONE in 8ms (0 errors, 0 warnings)
    Processing ValidationErrorRenderer.php [PHP => 401 tokens in 41 lines]... DONE in 15ms (0 errors, 0 warnings)
    Processing ValidateConfigCommand.php [PHP => 778 tokens in 98 lines]... DONE in 9ms (0 errors, 0 warnings)
    Changing into directory /vagrant/cli
    Processing FundraisingCli.php [PHP => 271 tokens in 39 lines]... DONE in 4ms (0 errors, 0 warnings)
    Processing RenderMailTemplatesCommand.php [PHP => 1297 tokens in 156 lines]... DONE in 25ms (0 errors, 0 warnings)
    Time: 440ms; Memory: 4Mb

File: tests/Data/mail_templates.php (hunk 3 above)
Comment (KaiNissen) [inline tests/Data/mail_templates.php]: No objections. I looked it up which led me to that assumption.

**Comment by wiese** [comment]
@KaiNissen Unfortunately there are no clear AC. My opinion: If we both think it is helping us in getting preview of the mails, it's done.

**Comment by wiese** [comment]
I think it's good to go now (despite the branch name).

File: cli/RenderMailTemplatesCommand.php (hunk 2 above)
Comment (KaiNissen) [inline cli/RenderMailTemplatesCommand.php]: I got used to have space between keywords and brackets. But I agree, as long as phpcs does not complain, there is no issue.

**Pull request merged**